src/compiler/checker.ts

```diff
@@ -4854,12 +4838,27 @@ namespace ts {
                             typeName,
                             typeArguments);
                     }
+
+                    function updateEntityTypeArguments(name: EntityName, typeArguments: NodeArray<TypeNode> | undefined) {
+                        if (isIdentifier(name)) {
+                            return factory.updateIdentifier(name, typeArguments);
+                        }
+                        if (isStringLiteralLike(name)) {
+                            return name; // TODO ?

```

Comment:
Can a string entity name have type arguments?

```ts
// @Filename: thing.js
/** @template T */
class Thing {
  /** @param {T} x */
  foo(x) {}
}
exports["a thing"] = Thing;

// @Filename: index.ts
type Thing<T> = import('./thing')."a thing"<T>
```

?

Reply:
Sort-of yes - those type arguments are attached to the import type (or type reference) itself, though. The type arguments sometimes attached to identifiers are provided for the language service in some cases to provide detailed instantiation information for qualified names (like when we return strings like `NS.Cls<T>.B`) - currently a `StringLiteralLike` doesn't support appending those, so I'm just omitting them, rather than complicate the types. To be honest, I'd prefer to move them into the `QualifiedName` structure instead of having them on `Identifier` (and now potentially string literals) if we wanted to uniformly keep the behavior.